locator/token_metadata.cc

```diff
@@ -904,11 +913,15 @@ future<> token_metadata_impl::update_pending_ranges(
 }
 
 size_t token_metadata_impl::count_normal_token_owners() const {
+    return _nr_normal_token_owners;
+}
+
+void token_metadata_impl::update_normal_token_owners() {
     std::set<inet_address> eps;
     for (auto [t, ep]: _token_to_endpoint_map) {
         eps.insert(ep);
     }
-    return eps.size();
+    _nr_normal_token_owners = eps.size();
```

Comment:
Traversing `_token_to_endpoint_map` (again e.g. in `update_normal_tokens`) is pretty heavy by itself.

If we keep a `std::set<inet_address>` of the normal token owners as a member and keep it up-to-date in the various methods that affect it, `count_normal_token_owners` can just return `normal_token_eps.size()`.

Reply:
Looping through _token_to_endpoint_map to get the size is what we do in count_normal_token_owners before this patch. Adding another std::set<inet_address> objects to track the normal nodes can indeed speed up the calculation. The downside is that we have to keep same information in two different places which we would like to avoid if possible.

Reply:
It's not the same information since the unique set of endpoints is not available directly from the multi-map, it needs to be calculated. So it's more like having multiple indexes over the same data rather than copying the data itself IMO.

Reply:
https://github.com/scylladb/scylla/issues/11128

Reply:
Possible silly suggestion but perhaps we need some custom data_structure that keeps both indexes up to date (possibly 
we need more than two prisms over the same data?)

Reply:
@eliransin `token_metadata` IS practically that custom data structure.
I don't think we need to implement a sophisticated ADT that does both indexes at the low level.